api4/user.go

```diff
@@ -1131,6 +1151,37 @@ func updateUser(c *Context, w http.ResponseWriter, r *http.Request) {
 	w.Write([]byte(ruser.ToJson()))
 }
 
+// checkProviderAttributes returns the empty string if the patch can be applied without
+// overriding attributes set by the user's login provider; otherwise, the name of the offending
+// field is returned.
+func checkProviderAttributes(a app.AppIface, ouser *model.User, patch *model.UserPatch) string {
+	tryingToChange := func(patchValue *string, userValue *string) bool {
+		return patchValue != nil && *patchValue != *userValue
+	}
+	LdapSettings := &a.Config().LdapSettings
+	SamlSettings := &a.Config().SamlSettings
+
+	conflictField := ""
+	if a.Ldap() != nil &&
+		(ouser.IsLDAPUser() || (ouser.IsSAMLUser() && *SamlSettings.EnableSyncWithLdap)) {
+		conflictField = a.Ldap().CheckProviderAttributes(LdapSettings, ouser, patch)
+	} else if a.Saml() != nil && ouser.IsSAMLUser() {
+		conflictField = a.Saml().CheckProviderAttributes(SamlSettings, ouser, patch)
+	} else if ouser.IsOAuthUser() {
+		if tryingToChange(patch.FirstName, &ouser.FirstName) ||
+			tryingToChange(patch.LastName, &ouser.LastName) {
+			conflictField = "full name"
+		}
+	}
```

Comment:
We can avoid nested if statements here
```suggestion
	} else if ouser.IsOAuthUser() && (tryingToChange(patch.FirstName, &ouser.FirstName) || tryingToChange(patch.LastName, &ouser.LastName)) {
			conflictField = "full name"
	}
```

Reply:
I'd prefer to keep it nested because that way we have one if clause for each authentication type. If we decide to add another clause for a new authentication type in the future, OAuth users wouldn't have to check the new clause.